broker/server_apis.go

```diff
@@ -0,0 +1,246 @@
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Failed to verify provided token"})
+		return
+	}
+	if !ok {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Authorization denied"})
+	}
+
+	req, err := handleRetrieve(ctx, ginCtx, originReq.Prefix, originReq.Origin, timeoutVal)
+	if errors.Is(err, errRetrieveTimeout) {
+		ginCtx.JSON(http.StatusOK, gin.H{"status": "timeout", "request": gin.H{}})
+		return
+	} else if err != nil {
+		ginCtx.JSON(http.StatusInternalServerError, brokerErrResp{Status: "error", Msg: "Failure when retrieving requests for this origin"})
+		return
+	}
+
+	ginCtx.JSON(http.StatusOK, gin.H{"status": "ok", "request": req})
+}
+
+func reverseRequest(ctx context.Context, ginCtx *gin.Context) {
+	timeoutStr := "5s"
+	if val := ginCtx.Request.Header.Get("X-Pelican-Timeout"); val != "" {
+		timeoutStr = val
+	}
+
+	timeoutVal, err := time.ParseDuration(timeoutStr)
+	if err != nil {
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Failed to parse X-Pelican-Timeout header to a duration (example: 5s)"})
+		return
+	}
+
+	token := ginCtx.Request.Header.Get("Authorization")
+	token, hasPrefix := strings.CutPrefix(token, "Bearer ")
+	if !hasPrefix {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Bearer authorization required for callback"})
+		return
+	}
+
+	hostname, err := getCacheHostnameFromToken([]byte(token))
+	if err != nil {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Failed to determine issuer: " + err.Error()})
+		return
+	}
+
+	ok, err := verifyToken(ctx, token, "/caches/"+hostname, param.Server_ExternalWebUrl.GetString(), token_scopes.Broker_Reverse)
+	if err != nil {
+		log.Errorln("Failed to verify token for cache reversal request:", err)
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Failed to verify provided token"})
+		return
+	}
+	if !ok {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Authorization denied"})
+	}
+
+	reversalReq := reversalRequest{}
+	if err := ginCtx.Bind(&reversalReq); err != nil {
+		ginCtx.JSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Failed to parse the cache's reversal request"})
+		ginCtx.Abort()
+		return
+	}
+	if reversalReq.OriginName == "" {
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Missing 'origin' parameter in request"})
+		return
+	}
+	if reversalReq.Prefix == "" {
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Missing 'prefix' parameter in request"})
+		return
+	}
+
+	if err = handleRequest(ctx, reversalReq.OriginName, reversalReq, timeoutVal); errors.Is(err, errRequestTimeout) {
+		ginCtx.JSON(http.StatusInternalServerError, gin.H{"status": "error", "msg": "Timeout when waiting for origin callback"})
```

Comment:
```suggestion
		ginCtx.JSON(http.StatusInternalServerError, gin.H{"status": "timeout", "msg": "Timeout when waiting for origin callback"})
```
Also, we return 200 with `status: timeout` in `retrieveRequest`, do we want to keep consistency here?

Reply:
This is purposeful.  In `retrieveRequest`, it is not considered an error if there are no requests to retrieve -- an "empty list" is fine.

Here, we asked the origin to perform a callback and it never did.